### ISSUE TITLE
zombie kube sessions never exit

### ISSUE THREAD

**Issue by programmerq** [gravitational/teleport#17848]
### Expected behavior:

When a user is disconnected from a kube session, it should not continue indefinitely as an active session.

### Current behavior:

A user is connected to a kube session, such as `kubectl exec -it pod/mypod -- bash`. Their connection is interrupted in a way that their client kubectl nor os is able to send any RST/FIN packet to the proxy (network cable unplug, sudden power failure of client machine or other network equipment, OpenVPN disconnect, etc). Their session still shows up as an active session in the Teleport console. It is never cleaned up, and the recording is not uploaded. Additionally, the bash process still exists in the container.

This was observed in a cluster with proxy-sync session recording mode.

### Bug details:
- Teleport version: 10.3.3

### Recreation steps

The reporting user runs Teleport in proxy-sync recording mode. They created kubernetes manifests using the teleport-cluster helm chart, but have customized the manifest files directly since. They use dynamodb for storage and events, s3 for session recordings, and every pod in the deployment is auth+proxy+kubernetes_service. They only have the one kubernetes cluster in this environment.

<details>
<summary>teleport.yaml with proxy-sync recording mode.</summary>

```yaml
auth_service:
  authentication:
    local_auth: "true"
    second_factor: "off"
    #type: saml
  client_idle_timeout: 30m
  client_idle_timeout_message: Terminating session as no activities.
  cluster_name: proxysynclab
  enabled: true
  license_file: /var/lib/teleport/license.pem
  proxy_listener_mode: multiplex
  session_recording: proxy-sync
  web_idle_timeout: 10m
kubernetes_service:
  enabled: true
  kube_cluster_name: mykubecluster
  labels:
    environment_class: lab
  listen_addr: 0.0.0.0:3027
proxy_service:
  enabled: true
  https_keypairs:
    - cert_file: /etc/teleport-tls/tls.crt
      key_file: /etc/teleport-tls/tls.key
  kube_listen_addr: 0.0.0.0:3026
  mysql_listen_addr: 0.0.0.0:3036
  public_addr: teleport.example.com:443
  web_listen_addr: 0.0.0.0:443
ssh_service:
  enabled: true
teleport:
  log:
    format:
      extra_fields:
        - timestamp
        - level
        - component
        - caller
      output: text
    output: stdout
    severity: DEBUG
  storage:
    audit_events_uri:
      - dynamodb://teleport.example.com-audit
      - stdout://
    audit_sessions_uri: s3://teleport.example.com
    continuous_backups: false
    region: us-east-2
    table_name: teleport.example.com-backend
    type: dynamodb
version: 2
```
</details>

Once the lab was running, I was able to reproduce the condition by starting a kubectl exec session, and then use iptables to completely block the flow of packets on the already established connection. kubectl itself exited when it could no longer communicate on the established connection. The teleport web interface still shows the session as active the next day. Load balancer timeouts in front of the lab are set to 600 seconds.

![image](https://user-images.githubusercontent.com/37278/198085914-1310a8f5-63ae-4a56-8691-a68ebb88ec33.png)

<details>
<summary>Click for details on the iptables blocking</summary>

I utilized a docker container to make it easy to implement an iptables rule that only affected my containerized kubectl exec procedure.

```
docker run --name kubetest -it --entrypoint='' quay.io/gravitational/teleport-ent:10.3.3 bash --login
root@ba1bf16096d0:/# apt -y install net-tools curl
root@ba1bf16096d0:/# curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl" && chmod +x kubectl && mv kubectl /usr/local/bin/kubectl
root@ba1bf16096d0:/# tsh login --proxy=teleport.example.com:443 --user=jeff proxysynclab
root@ba1bf16096d0:/# export KUBECONFIG=${HOME?}/teleport-kubeconfig.yaml
root@ba1bf16096d0:/# tsh kube login mykubecluster
root@ba1bf16096d0:/# kubectl get ns
root@ba1bf16096d0:/# kubectl get pods
root@ba1bf16096d0:/# kubectl exec -it mypod -- bash --login
```

Then in another terminal, I can exec into the docker container as a privileged user and add my own iptables rule:

```
docker exec -it --priveleged kubetest bash --login
root@ba1bf16096d0:/# iptables -nL
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
root@ba1bf16096d0:/# iptables -I OUTPUT -p tcp --dport 443 -j DROP
root@ba1bf16096d0:/# iptables -nL
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
DROP       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443
root@ba1bf16096d0:/# iptables-save -c
# Generated by iptables-save v1.8.4 on Wed Oct 26 15:46:00 2022
*filter
:INPUT ACCEPT [18:3354]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
[52:11009] -A OUTPUT -p tcp -m tcp --dport 443 -j DROP
COMMIT
# Completed on Wed Oct 26 15:46:00 2022
```

Rerunning the last command shows the counts of the rules. As I typed in my kubectl exec session, that count went up. Finally, kubectl's connection timed out because no response packets ever made it back, and kubectl exited.

The same iptables rule could be used for any other number of environments. The privileged docker container is just a handy approach.

</details>

Running `ps faux` inside the pod shows that the original bash process is still running:

```
$ kubectl -n other exec -it mypod -- ps faux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root     31621  0.0  0.1   5896  2916 pts/3    Rs+  16:29   0:00 ps faux
root      1980  0.0  0.1   4244  3496 pts/0    Ss+  Oct25   0:00 bash --login
root         1  0.3  7.3 1618336 147056 ?      Ssl  Oct24   8:54 nginx -g daemon
```

The session will stay as active until the teleport auth/proxy/agent pod is cycled. This doesn't always result in a successful session.end and session upload event. It simply no longer shows up in active sessions.

Stopping the `mypod` pod will result in the session ending as well. The recording will be uploaded too.

Killing the bash process inside the container also causes the session to gracefully end.

- Debug logs 
[waitlogs.txt.gz](https://github.com/gravitational/teleport/files/9872009/waitlogs.txt.gz) (entire logs for the lab over the course of a day, including a USR1 signal sent while the session was stuck. It has audit events mixed in to the log stream as well.

**Comment by tunguyen9889** [comment]
Hi @programmerq, will the fix be backport'ed to version 10? Or it's only released for version 11?

**Comment by tigrato** [comment]
@tunguyen9889 
The fix was already backported to V10
#18686

**Comment by tunguyen9889** [comment]
I just saw it's just released 4 hours ago (10.3.9). I will try that version in our clusters and verify.
